src/main/flight/pid.c

```diff
@@ -981,44 +904,39 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, timeUs_t currentTim
     rpmFilterUpdate();
 #endif
 
-const bool newRcFrame = getShouldUpdateFeedforward();
-
     // ----------PID controller----------
     for (int axis = FD_ROLL; axis <= FD_YAW; ++axis) {
 
         float currentPidSetpoint = getSetpointRate(axis);
-        float rawSetpoint = getRawSetpoint(axis);
-        bool rawSetpointIsSmoothed = false;
         if (pidRuntime.maxVelocity[axis]) {
             currentPidSetpoint = accelerationLimit(axis, currentPidSetpoint);
         }
         // Yaw control is GYRO based, direct sticks control is applied to rate PID
         // When Race Mode is active PITCH control is also GYRO based in level or horizon mode
 #if defined(USE_ACC)
-        // earth reference yaw by attenuating yaw rate at higher angles
-        if (axis == FD_YAW && pidRuntime.angleEarthRef) {
-            pidRuntime.angleYawSetpoint = currentPidSetpoint;
+        pidRuntime.axisInAngleMode[axis] = false;
+        if (axis < FD_YAW) {
+            if (levelMode == LEVEL_MODE_RP || (levelMode == LEVEL_MODE_R && axis == FD_ROLL)) {
+                pidRuntime.axisInAngleMode[axis] = true;
+                currentPidSetpoint = pidLevel(axis, pidProfile, angleTrim, currentPidSetpoint, horizonLevelStrength);
+            }
+        } else { // yaw axis only
             if (levelMode == LEVEL_MODE_RP) {
-                float maxAngleTargetAbs = fmaxf( fabsf(pidRuntime.angleTarget[FD_ROLL]), fabsf(pidRuntime.angleTarget[FD_PITCH]) );
-                maxAngleTargetAbs *= pidRuntime.angleEarthRef;
-                if (FLIGHT_MODE(HORIZON_MODE)) {
-                    // reduce yaw compensation when Horizon uses less levelling
-                    maxAngleTargetAbs *= horizonLevelStrength;
+                // if earth referencing is requested, attenuate yaw axis setpoint when pitched or rolled
+                // and send yawSetpoint to Angle code to modulate pitch and roll
+                if (pidRuntime.angleEarthRef) {
+                    pidRuntime.angleYawSetpoint = currentPidSetpoint;
+                    float maxAngleTargetAbs = fmaxf( fabsf(pidRuntime.angleTarget[FD_ROLL]), fabsf(pidRuntime.angleTarget[FD_PITCH]) );
+                    maxAngleTargetAbs *= pidRuntime.angleEarthRef; // allow user control
+                    if (FLIGHT_MODE(HORIZON_MODE)) {
+                        maxAngleTargetAbs *= horizonLevelStrength; // reduce compensation while Horizon uses less levelling
+                    }
+                    float attenuateYawSetpoint = cos_approx(DEGREES_TO_RADIANS(maxAngleTargetAbs));
+                    currentPidSetpoint *= attenuateYawSetpoint;
```

Comment:
```suggestion
                    currentPidSetpoint *= cos_approx(DEGREES_TO_RADIANS(maxAngleTargetAbs));  // Attenuate yaw setpoint
```

Reply:
I measured both options, they took exactly the same time - 86 cycles from before declaring maxAngleTargetAbs, to the debug.  Of those 86 cycles, 79 come from the cosApprox function.
I'll implement your suggestion.